SSA/Core/Framework.lean

```diff
@@ -220,282 +590,717 @@ args, since there now are equation lemmas for it.
 #eval Lean.Meta.getEqnsFor? ``Expr.denote
 #eval Lean.Meta.getEqnsFor? ``Com.denote
 
-def Lets.denote : Lets Op Γ₁ Γ₂ → Valuation Γ₁ → Valuation Γ₂
-  | .nil => id
-  | .lete e body => fun ll t v => by
-    cases v using Var.casesOn with
-    | last =>
-      apply body.denote
-      apply e.denote
-      exact ll
-    | toSnoc v =>
-      exact e.denote ll v
+end Unfoldings
+
+/-! simp-lemmas about `denote` functions -/
+section Lemmas
+
+@[simp] lemma Com.denote_ret {eff : EffectKind} (Γ : Ctxt Ty) (x : Γ.Var t) [Monad m] :
+    (Com.ret x : Com Op Γ eff t).denote = fun Γv => return (Γv x) := by
+  funext Γv
+  simp [denote]
+
+@[simp] lemma Com.denote_lete [LawfulMonad m] {e : Expr Op Γ eff α} :
+    (Com.lete e body).denote =
+    fun Γv => (e.denote (m := m) Γv) >>= (fun v => body.denote (Γv.snoc v)) := by
+  funext Γv
+  cases eff <;> simp [denote]
+
+@[simp] lemma Com.denoteLets_lete (e : Expr Op Γ eff t) (body : Com Op _ eff u) [LawfulMonad m] :
+    (Com.lete e body).denoteLets =
+        (fun V => e.denote V >>= fun Ve => body.denoteLets (V.snoc Ve)) := by
+  funext V
+  cases eff <;> simp [denoteLets, bind_pure]
+
+@[simp] lemma Com.denoteImpure_ret [Monad m] [OpDenote Op Ty m] {Γ : Ctxt Ty} (x : Γ.Var t) :
+  (Com.ret (Op := Op) (eff := eff) x).denoteImpure = fun Γv => return (Γv x) := rfl
+
+@[simp] lemma Com.denoteImpure_body [Monad m] [OpDenote Op Ty m] {Γ : Ctxt Ty}
+    (e : Expr Op Γ eff te) (body : Com Op (Γ.snoc te) eff tbody) :
+  (Com.lete e body).denoteImpure =
+  fun Γv => e.denoteImpure Γv >>= fun x => body.denote (Γv.snoc x) := rfl
+
+@[simp] lemma Lets.denote_nil {Γ : Ctxt Ty} :
+    (Lets.nil : Lets Op Γ eff Γ).denote = (return ·) := by
+  funext; simp [denote]
+-- TODO: make `[simp]`
```

Comment:
I don't understand this TODO. What's preventing this from being `@[simp]`?

Reply:
As the author of the comment, neither do I, so I nuked it.